src/BucketTesting/Validation/CampaignUtilizationValidator.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\BucketTesting\Validation;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\NodeFinder;
+use PhpParser\ParserFactory;
+use Symfony\Component\Filesystem\Exception\FileNotFoundException;
+use WMDE\Fundraising\Frontend\BucketTesting\Campaign;
+use WMDE\Fundraising\Frontend\BucketTesting\CampaignCollection;
+
+/**
+ * @license GNU GPL v2+
+ */
+class CampaignUtilizationValidator {
+
+	/** @see \WMDE\Fundraising\Frontend\Factories\ChoiceFactory */
+	const CHOICE_FACTORY_LOCATION = 'src/Factories/ChoiceFactory.php';
+	/** @see \WMDE\Fundraising\Frontend\BucketTesting\FeatureToggle::featureIsActive */
+	const FEATURE_TOGGLE_METHOD_NAME = 'featureIsActive';
+
+	private $hasValidated = false;
+	private $campaignCollection;
+	private $errorLogger;
+	private $ignoredBuckets = [];
+
+	public function __construct( CampaignCollection $campaignCollection, array $ignoredBuckets, ValidationErrorLogger $errorLogger ) {
+		$this->campaignCollection = $campaignCollection;
+		$this->errorLogger = $errorLogger;
+		$this->ignoredBuckets = $ignoredBuckets;
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errorLogger->getErrors();
+	}
+
+	private function validate(): void {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		$featureToggleChecks = $this->getFeatureToggleChecks();
+		if ( $duplicates = $this->getDuplicates( $featureToggleChecks ) ) {
+			foreach ( $duplicates as $duplicate ) {
+				$this->errorLogger->addError(
+					'ChoiceFactory contains multiple checks for feature toggle: ' . $duplicate
+				);
+			}
+			return;
+		}
```

Comment:
If a campaign necessitates multiple code changes (e.g. a change where both presentation and database model need to be changed), this duplicate check would fail. I still find the check useful and necessary and would leave it in, if the situation ever arises, we'll figure something out.

For now, the `return` is not strictly necessary, I'd go on checking even if there are duplicates. Maybe you need to call `array_uniq` for the following logic, but that's ok.

Also you move the `if` section to a private method, e.g. `$this->checkForDuplicateFeatureToggles( $featureToggleChecks )`?